src/Microsoft.OData.Core/Evaluation/ODataResourceMetadataContext.cs

```diff
@@ -267,13 +270,38 @@ private static object GetPrimitiveOrEnumPropertyValue(string entityTypeName, ODa
         /// </summary>
         /// <param name="keyProperties">Key properties of the resource.</param>
         /// <param name="actualEntityTypeName">The entity type name of the resource.</param>
-        private static void ValidateEntityTypeHasKeyProperties(KeyValuePair<string, object>[] keyProperties, string actualEntityTypeName)
+        /// <param name="throwOnError">Whether to throw if validation fails.</param>
+        /// <returns>True, if validation succeeds, or false if validation fails.</returns>
+        private static bool ValidateEntityTypeHasKeyProperties(KeyValuePair<string, object>[] keyProperties, string actualEntityTypeName, bool throwOnError)
         {
-            Debug.Assert(keyProperties != null, "keyProperties != null");
             if (keyProperties == null || keyProperties.Length == 0)
             {
-                throw new ODataException(Strings.ODataResourceMetadataContext_EntityTypeWithNoKeyProperties(actualEntityTypeName));
+                if (throwOnError)
+                {
+                    throw new ODataException(Strings.ODataResourceMetadataContext_EntityTypeWithNoKeyProperties(actualEntityTypeName));
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            for (int keyProperty = 0; keyProperty < keyProperties.Length; keyProperty++)
+            {
+                if (keyProperties[keyProperty].Value == null || (keyProperties[keyProperty].Value is ODataValue && !(keyProperties[keyProperty].Value is ODataEnumValue)))

```

Comment:
It appears that the compiler is not smart enough to optimize the multiple indexing operations (probably because nothing guarantees that there are no side effects):

[Multiple indexes](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2WMALJgLIAUAysAE5gB2A5gNoC6mAhk3YBnAJSFipSQDMA9k0y02wTGEwBeTAAYA3CswAefoKEA6ADIBTDsAAWugNT2wIiZJJEUbryqkKBwzjBeAEINVgBXCAgXT29JDzi4mAQATlp/IUDuEW1XRIBfPMlC2JIS/KA===)
[Storing the indexed value](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2WMALJgLIAUAysAE5gB2A5gNoC6mAhk3YBnAJSFipSQDMA9k0y02wTGEwBeTAAYA3CswAefoKEA6ADIBTDsAAWugNT2wIiZJJEUbr5gBuAo+zqAUKcYNzart4kYFIKAoEAhBqsAK4QEC6eUaQe2d4wCACctPGiEVl5AL6RktUVmHWVQA=)

In the first example, the IL indicates that the indexing is performed a second time for the `WriteLine` call. The JITer might be able to optimize this, but there's really no way to know for sure. It would probably make the most sense to store `keyProperties[keyProperty]` in a variable scoped to the loop so that we only index the one time, instead of 4 times.

Reply:
Good catch -- thanks!